{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/vvithurshan/Homology-Modelling-for-Catnap-Dataset/blob/main/homology_modelling_catnap.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "KuboB_kaaZt7"
      },
      "outputs": [],
      "source": [
        "import mdtraj as md\n",
        "from Bio import pairwise2\n",
        "import MDAnalysis as mda\n",
        "import numpy as np\n",
        "from abnumber import Chain\n",
        "from Bio.Align import substitution_matrices\n",
        "import pandas as pd\n",
        "import shutil\n",
        "from modeller.automodel import *\n",
        "from modeller import *\n",
        "import os"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Z4BAWbTkaZuB"
      },
      "outputs": [],
      "source": [
        "Virus_dict = {}\n",
        "with open(\"./CATNAP_DATABASE/Virus_aligned_removed.fasta\", 'r') as f:\n",
        "  for line in f:\n",
        "    if line.startswith(\">\"):\n",
        "      key = line[1:].split('.')[-2]\n",
        "    else:\n",
        "      Virus_dict[key] = line[24:510]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "UQukR84JaZuC"
      },
      "outputs": [],
      "source": [
        "with open('./CATNAP_DATABASE/heavy_seqs_aa.fasta', 'r') as f:\n",
        "    heavy_dict = {}\n",
        "    current_id = \"\"\n",
        "    for line in f:\n",
        "        line = line.strip()\n",
        "        if line.startswith(\">\"):\n",
        "            current_id = line[1:].split('_')[0]\n",
        "            heavy_dict[current_id] = \"\"\n",
        "        else:\n",
        "            heavy_dict[current_id] += line"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "BorEVTL6aZuD"
      },
      "outputs": [],
      "source": [
        "with open('./CATNAP_DATABASE/light_seqs_aa.fasta', 'r') as f:\n",
        "    light_dict = {}\n",
        "    current_id = \"\"\n",
        "    for line in f:\n",
        "        line = line.strip()\n",
        "        if line.startswith(\">\"):\n",
        "            current_id = line[1:].split('_')[0]\n",
        "            light_dict[current_id] = \"\"\n",
        "        else:\n",
        "            light_dict[current_id] += line"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "NZmUMpLMaZuE"
      },
      "outputs": [],
      "source": [
        "def SeqAlignWrite(target_seq, tem_Ab, tem_An):\n",
        "    tem_Ab_pdb = \"{}.pdb\".format(tem_Ab)\n",
        "    tem_An_pdb = \"{}.pdb\".format(tem_An)\n",
        "    tem_Ab_pdb_path = \"../../../Template_PDB/\" + tem_Ab_pdb\n",
        "    tem_An_pdb_path = \"../../../Template_PDB/\" + tem_An_pdb\n",
        "\n",
        "    ## copy template pdb\n",
        "    shutil.copy(tem_Ab_pdb_path, os.getcwd())\n",
        "    shutil.copy(tem_An_pdb_path, os.getcwd())\n",
        "\n",
        "    ##chains to fasta\n",
        "    tem_ab = mda.Universe(tem_Ab_pdb)\n",
        "    protein_atoms_ab = tem_ab.select_atoms(\"protein\") \n",
        "\n",
        "    template_Ab_chain_ids = [i.segid for i in protein_atoms_ab.segments]\n",
        "\n",
        "    tem_an = mda.Universe(tem_An_pdb)\n",
        "    protein_atoms_an = tem_an.select_atoms(\"protein\") \n",
        "    template_An_chain_ids = [i.segid for i in protein_atoms_an.segments]\n",
        "\n",
        "    traj_ab = md.load(tem_Ab_pdb).topology\n",
        "    template_Ab_chainsToFasta = [traj_ab.to_fasta(i) for i in range(len(template_Ab_chain_ids))]\n",
        "\n",
        "    traj_an = md.load(tem_An_pdb).topology\n",
        "    template_An_chainsToFasta = [traj_an.to_fasta(i) for i in range(len(template_An_chain_ids))]\n",
        "\n",
        "    ## is first a heavy_chain?\n",
        "    Is_heavy = Chain(template_Ab_chainsToFasta[0] , scheme='chothia', cdr_definition='chothia')\n",
        "\n",
        "    if not Is_heavy.is_heavy_chain():\n",
        "        target_seq[0], target_seq[1] = target_seq[1], target_seq[0]\n",
        "    \n",
        "    ## row = len of template chain ids\n",
        "    ## col = len of tar_seq\n",
        "\n",
        "    row = len(template_Ab_chain_ids) + len(template_An_chain_ids)\n",
        "    col = len(target_seq)\n",
        "\n",
        "    score_matrix = np.zeros((row, col))\n",
        "    ## fill alignment score\n",
        "\n",
        "    for i, seq_Ab in enumerate(template_Ab_chainsToFasta):\n",
        "        for j, seq_tar in enumerate(target_seq):\n",
        "            alignments = pairwise2.align.globalxx(seq_Ab, seq_tar)\n",
        "            score_matrix[i][j] = alignments[0].score\n",
        "            # print(alignments[0].score)\n",
        "\n",
        "\n",
        "    for i, seq_An in enumerate(template_An_chainsToFasta):\n",
        "        for j, seq_tar in enumerate(target_seq):\n",
        "            alignments = pairwise2.align.globalxx(seq_An, seq_tar)\n",
        "            score_matrix[i+len(template_Ab_chainsToFasta)][j] = alignments[0].score\n",
        "            # print(alignments[0].score)\n",
        "\n",
        "    ## \n",
        "    Target = []\n",
        "    Template_AB = []\n",
        "    Template_AN = []\n",
        "    \n",
        "    ## alignment parameters\n",
        "    matrix = substitution_matrices.load(\"BLOSUM62\")\n",
        "    open_gap_penalty = -10\n",
        "    extend_gap_penalty = -1\n",
        "   \n",
        "    ## do the alignment\n",
        "    for i in range(col):\n",
        "        max_val = np.max(score_matrix[:,i])\n",
        "        max_id = np.where(score_matrix[:,i] == max_val)\n",
        "        max_id = int(max_id[0])\n",
        "        if max_id < len(template_Ab_chainsToFasta):\n",
        "            alignments = pairwise2.align.localds(template_Ab_chainsToFasta[max_id], target_seq[i], matrix, open_gap_penalty, extend_gap_penalty)\n",
        "            Template_AB.append(alignments[0].seqA)\n",
        "            Target.append(alignments[0].seqB)\n",
        "            Template_AN.append((\"-\" * len(alignments[0].seqA)))\n",
        "        else:\n",
        "            alignments = pairwise2.align.localds(template_An_chainsToFasta[max_id - len(template_Ab_chainsToFasta)], target_seq[i], matrix,  open_gap_penalty, extend_gap_penalty)\n",
        "            Template_AN.append(alignments[0].seqA)\n",
        "            Target.append(alignments[0].seqB)\n",
        "            Template_AB.append((\"-\" * len(alignments[0].seqA)))  \n",
        "\n",
        "    with open(\"sequence.ali\", \"w\") as fp:\n",
        "        ##target\n",
        "        fp.write(\">P1;complex_seq\\n\")\n",
        "        fp.write(\"sequence::.:.:.:.::::\\n\")\n",
        "        tar_str = \"/\".join(Target)\n",
        "        fp.write(tar_str + \"*\\n\")\n",
        "        ##ab\n",
        "        fp.write(f\">P1;\" + tem_Ab_pdb + \"\\n\")\n",
        "        fp.write(f\"structure:\" + tem_Ab_pdb + \":FIRST:@:LAST:.::::\\n\")\n",
        "        tem_ab_str = \"/\".join(Template_AB)\n",
        "        fp.write(tem_ab_str+\"*\\n\")\n",
        "        ##an\n",
        "        fp.write(f\">P1;\" + tem_An_pdb + \"\\n\")\n",
        "        fp.write(f\"structure:\" + tem_An_pdb + \":FIRST:@:LAST:.::::\\n\")\n",
        "        tem_an_str = \"/\".join(Template_AN)\n",
        "        fp.write(tem_an_str+\"*\\n\")\n",
        "    \n",
        "    return (tem_Ab_pdb, tem_An_pdb)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "scrolled": false,
        "id": "ga0A352RaZuF"
      },
      "outputs": [],
      "source": [
        "import os \n",
        "import subprocess\n",
        "\n",
        "\n",
        "## change the directory\n",
        "def ChangeDir(Antibody, Antigen):\n",
        "\n",
        "    if not os.path.exists(Antibody):\n",
        "        os.makedirs(Antibody)\n",
        "    os.chdir(Antibody)\n",
        "\n",
        "    folder_name = \"{}_{}\".format(Antibody,Antigen)\n",
        "    if not os.path.exists(folder_name):\n",
        "        os.makedirs(folder_name)\n",
        "    os.chdir(folder_name)\n",
        "\n",
        "def Modeller(templates_tuple, output):\n",
        "\n",
        "    ## modeller \n",
        "    env = environ()\n",
        "    print(templates_tuple)\n",
        "    a = automodel(env, alnfile='sequence.ali',\n",
        "                knowns= templates_tuple, sequence='complex_seq')\n",
        "\n",
        "    a.starting_model = 0\n",
        "    a.ending_model = 0\n",
        "    # a.auto_align()\n",
        "\n",
        "    a.md_level = None\n",
        "    # a.repeat_optimization = 2\n",
        "    # a.md_level = refine.slow\n",
        "    # a.max_var_iterations = 300\n",
        "    # a.repeat_optimization = 2\n",
        "    # a.max_molpdf = 1e6\n",
        "    # a.read_template(template_path)\n",
        "    a.make()\n",
        "\n",
        "    model_output = \"{}.pdb\".format(output)\n",
        "    shutil.copy(\"complex_seq.B99990000.pdb\", model_output)\n",
        "    name = 'vvithurshan'\n",
        "    new_name = 'V _VITHURSHAN'\n",
        "    subprocess.run(['sed', '-i', f's/{name}/{new_name}/g', model_output])\n",
        "\n",
        "    storing_location = \"/home/vvithurshan/0014/vvarenthirarajah/Documents/ML-Antibody-Antigen-2023/Homology Modeling/CATNAP/Homology_modelling_catnap/PDB_CATNAP\"\n",
        "    shutil.copy(model_output, storing_location)\n",
        "    return True\n",
        "\n",
        "\n",
        "def main(Antibody, Antigen, Target_sequence):\n",
        "    ## change dir\n",
        "    ChangeDir(Antibody, Antigen)\n",
        "\n",
        "    ##temple for ab and template for an\n",
        "    df_tem = pd.read_csv(\"../../../Ab_An_Template.csv\")\n",
        "    tem_Ab = df_tem[df_tem['Antibody'] == Antibody]['Ab_template'].iloc[0]\n",
        "    tem_An = df_tem[df_tem['Antibody'] == Antibody]['An_template'].iloc[0]\n",
        "\n",
        "    templates_tuple = SeqAlignWrite(target_seq = Target_sequence, tem_Ab = tem_Ab, tem_An = tem_An)\n",
        "\n",
        "    model_out = \"{}__{}\".format(Antibody, Antigen)\n",
        "\n",
        "    model = Modeller(templates_tuple = templates_tuple, output = model_out)\n",
        "\n",
        "    if model:\n",
        "        return model_out\n",
        "    else:\n",
        "        return 0\n",
        "if __name__ == \"__main__\":\n",
        "    ## list of antibodies to be modelled\n",
        "    os.chdir('/home/vvithurshan/0014/vvarenthirarajah/Documents/ML-Antibody-Antigen-2023/Homology Modeling/CATNAP/Homology_modelling_catnap')\n",
        "    df = pd.read_csv(\"../DataBase/test/tst.csv\")\n",
        "    df = df[['Antibody', 'Antigen']]\n",
        "    list_abs = ['VRC01']\n",
        "    df_model = df[df['Antibody'].isin(list_abs)]\n",
        "    # df_model = df[df['Antibody'] == '12A21'] ## temp\n",
        "\n",
        "    os.chdir('./model/')\n",
        "    \n",
        "    for index, row in df_model.iterrows():\n",
        "        ab = row['Antibody']\n",
        "        an = row['Antigen']\n",
        "        tar_seq_list = []\n",
        "        heavy = heavy_dict[ab].replace(\"X\",\"\")\n",
        "        light = light_dict[ab].replace(\"X\",\"\")\n",
        "        tar_seq_list.append(heavy)\n",
        "        tar_seq_list.append(light)\n",
        "        try:\n",
        "            virus_seq = Virus_dict[an].replace(\"X\",\"\")\n",
        "            tar_seq_list.append(virus_seq)\n",
        "        except:\n",
        "            df_model.at[index, 'model'] = 0\n",
        "            continue\n",
        "        modeler_model =main(Antibody = ab, Antigen = an, Target_sequence = tar_seq_list)\n",
        "        df_model.at[index, 'model'] = modeler_model\n",
        "\n",
        "        os.chdir(\"../../\")\n",
        "    df_model.to_csv(\"./VRC01/VRC01.csv\")"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.9.15 | packaged by conda-forge | (main, Nov 22 2022, 08:45:29) \n[GCC 10.4.0]"
    },
    "vscode": {
      "interpreter": {
        "hash": "1b173bd2b8203f09e53d07427f25d579d159175d54051585125192f9427d86f4"
      }
    },
    "colab": {
      "provenance": [],
      "include_colab_link": true
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}